File: trunk/C#/PokerProtocol/LobbyTCPClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using EricUtility;
using EricUtility.Collections;
using EricUtility.Networking;
using EricUtility.Networking.Commands;
using PokerProtocol.Commands.Lobby;
using PokerWorld.Game;
using PokerProtocol.Commands.Lobby.Training;
using PokerProtocol.Commands.Lobby.Career;

namespace PokerProtocol
{
    public delegate void DisconnectDelegate();
    public class LobbyTCPClient : TCPCommunicator
    {
        public event DisconnectDelegate ServerLost = delegate{};
        protected string m_PlayerName;
        protected string m_ServerAddress;
        protected int m_ServerPort;

        public string PlayerName
        {
            get { return m_PlayerName; }
        }

        public string ServerAddress
        {
            get { return m_ServerAddress; }
        }

        public int ServerPort
        {
            get { return m_ServerPort; }
        }
        protected Dictionary<int, GameClient> m_Clients = new Dictionary<int, GameClient>();
        protected BlockingQueue<string> m_Incoming = new BlockingQueue<string>();

        public LobbyTCPClient(string serverAddress, int serverPort)
            : base()
        {
            m_ServerAddress = serverAddress;
            m_ServerPort = serverPort;
        }

        public bool Connect()
        {
            return base.Connect(m_ServerAddress, m_ServerPort);
        }

        public void LeaveTable(int idGame)
        {
            if (m_Clients.ContainsKey(idGame))
            {
                GameClient client = m_Clients[idGame];
                m_Clients.Remove(idGame);
                if (client != null)
                {
                    client.Disconnect();
                }
            }
        }

        protected StringTokenizer ReceiveCommand(string expected)
        {
            string s = m_Incoming.Dequeue();
            StringTokenizer token = new StringTokenizer(s, AbstractLobbyCommand.Delimitter);
            string commandName = token.NextToken();
            while (s != null && commandName != expected)
            {
                s = m_Incoming.Dequeue();
                token = new StringTokenizer(s, AbstractLobbyCommand.Delimitter);
                commandName = token.NextToken();
            }
            return token;
        }

        protected string Receive(StreamReader reader)
        {
            string line;
            try
            {
                line = reader.ReadLine();
                LogManager.Log(LogLevel.MessageLow, "LobbyTCPClient.Receive", "{0} RECV [{1}]", m_PlayerName, line);
            }
            catch
            {
                return null;
            }
            return line;
        }
        public override void OnReceiveCrashed(Exception e)
        {
            if (e is IOException)
            {
                LogManager.Log(LogLevel.Error, "LobbyTCPClient.OnReceiveCrashed", "Lobby lost connection with server");
                Disconnect();
            }
            else
                base.OnReceiveCrashed(e);
        }
        public override void OnSendCrashed(Exception e)
        {
            if (e is IOException)
            {
                LogManager.Log(LogLevel.Error, "LobbyTCPClient.OnReceiveCrashed", "Lobby lost connection with server");
                Disconnect();
            }
            else
                base.OnSendCrashed(e);
        }
        public void Send(StreamWriter writer, AbstractCommand command)
        {
            writer.WriteLine(command.Encode());
        }
        public void Send(AbstractCommand command)
        {
            base.Send(command.Encode());
        }
        public void Disconnect()
        {
            foreach (GameClient client in m_Clients.Values)
                client.Disconnect();
            m_Clients.Clear();
            if (IsConnected)
            {
                Send(new DisconnectCommand());
                Close();
            }
        }
        public GameClient FindClient(int noPort)
        {
            if (m_Clients.ContainsKey(noPort))
                return m_Clients[noPort];
            return null;
        }

        protected virtual int GetJoinedSeat(int p_noPort, string player)
        {
            JoinTableCommand command = new JoinTableCommand(p_noPort, player);
            Send(command);

            StringTokenizer token2 = ReceiveCommand(JoinTableResponse.COMMAND_NAME);
            if (!token2.HasMoreTokens())
                return -1;
            JoinTableResponse response2 = new JoinTableResponse(token2);
            return response2.NoSeat;
        }

        public GameClient JoinTable(int p_noPort, string p_tableName, IPokerViewer gui)
        {
            int noSeat = GetJoinedSeat(p_noPort, m_PlayerName);
            if (noSeat == -1)
            {
                LogManager.Log(LogLevel.MessageLow, "LobbyTCPClient.JoinTable", "Cannot sit at this table: {0}", p_tableName);
                return null;
            }

            GameClient client = new GameClient(noSeat, m_PlayerName, p_noPort);
            client.SendedSomething += new EventHandler<KeyEventArgs<string>>(client_SendedSomething);
            if (gui != null)
            {
                gui.SetGame(client, client.NoSeat);
                gui.Start();
            }
            client.Start();
            m_Clients.Add(p_noPort, client);
            return client;
        }

        protected void client_SendedSomething(object sender, KeyEventArgs<string> e)
        {
            GameClient client = (GameClient)sender;
            Send(new GameCommand(client.NoPort, e.Key));
        }

        public List<TupleTableInfoTraining> getListTrainingTables()
        {
            Send(new ListTableCommand(true));

            StringTokenizer token = ReceiveCommand(ListTableTrainingResponse.COMMAND_NAME);
            if (token.HasMoreTokens())
            {
                ListTableTrainingResponse response = new ListTableTrainingResponse(token);
                return response.Tables;
            }
            else
                return new List<TupleTableInfoTraining>();
        }

        public List<TupleTableInfoCareer> getListCareerTables()
        {
            Send(new ListTableCommand(false));

            StringTokenizer token = ReceiveCommand(ListTableCareerResponse.COMMAND_NAME);
            if (token.HasMoreTokens())
            {
                ListTableCareerResponse response = new ListTableCareerResponse(token);
                return response.Tables;
            }
            else
                return new List<TupleTableInfoCareer>();
        }
        protected override void Run()
        {
            while (IsConnected)
            {
                LogManager.Log(LogLevel.MessageVeryLow, "LobbyTCPClient.Run", "{0} IS WAITING", m_PlayerName);
                string line = Receive();
                if (line == null)
                {
                    ServerLost();
                    return;
                }
                LogManager.Log(LogLevel.MessageLow, "LobbyTCPClient.Run", "{0} RECV [{1}]", m_PlayerName, line);
                StringTokenizer token = new StringTokenizer(line, AbstractLobbyCommand.Delimitter);
                String commandName = token.NextToken();
                if (commandName == GameCommand.COMMAND_NAME)
                {
                    GameCommand c = new GameCommand(token);
                    int count = 0;
                    while (!m_Clients.ContainsKey(c.TableID) && (count++ < 5))
                        Thread.Sleep(100);
                    if (m_Clients.ContainsKey(c.TableID))
                        m_Clients[c.TableID].Incoming(c.Command);
                }
                else
                    m_Incoming.Enqueue(line);
            }
        }


        public int CreateCareerTable(string p_tableName, int p_bigBlind, int p_maxPlayers, int wtaPlayerAction, int wtaBoardDealed, int wtaPotWon, TypeBet limit)
        {
            Send(new CreateCareerTableCommand(p_tableName, p_bigBlind, p_maxPlayers, m_PlayerName, wtaPlayerAction, wtaBoardDealed, wtaPotWon, limit));

            StringTokenizer token = ReceiveCommand(CreateCareerTableResponse.COMMAND_NAME);
            if (token.HasMoreTokens())
            {
                CreateCareerTableResponse response = new CreateCareerTableResponse(token);
                return response.Port;
            }
            else
                return -1;
        }


        public int CreateTrainingTable(string p_tableName, int p_bigBlind, int p_maxPlayers, int wtaPlayerAction, int wtaBoardDealed, int wtaPotWon, TypeBet limit, int startingMoney)
        {
            Send(new CreateTrainingTableCommand(p_tableName, p_bigBlind, p_maxPlayers, m_PlayerName, wtaPlayerAction, wtaBoardDealed, wtaPotWon, limit, startingMoney));

            StringTokenizer token = ReceiveCommand(CreateTrainingTableResponse.COMMAND_NAME);
            if (token.HasMoreTokens())
            {
                CreateTrainingTableResponse response = new CreateTrainingTableResponse(token);
                return response.Port;
            }
            else
                return -1;
        }
    }
}

File: trunk/JAVA/PokerProtocol/src/bluffinmuffin/protocol/LobbyTCPClient.java
package bluffinmuffin.protocol;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.StringTokenizer;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

import bluffinmuffin.poker.IPokerViewer;
import bluffinmuffin.poker.entities.type.GameBetLimitType;
import bluffinmuffin.protocol.commands.AbstractCommand;
import bluffinmuffin.protocol.commands.DisconnectCommand;
import bluffinmuffin.protocol.commands.lobby.AbstractLobbyCommand;
import bluffinmuffin.protocol.commands.lobby.GameCommand;
import bluffinmuffin.protocol.commands.lobby.JoinTableCommand;
import bluffinmuffin.protocol.commands.lobby.JoinTableResponse;
import bluffinmuffin.protocol.commands.lobby.ListTableCommand;
import bluffinmuffin.protocol.commands.lobby.career.ListTableCareerResponse;
import bluffinmuffin.protocol.commands.lobby.training.CreateTrainingTableCommand;
import bluffinmuffin.protocol.commands.lobby.training.CreateTrainingTableResponse;
import bluffinmuffin.protocol.commands.lobby.training.IdentifyCommand;
import bluffinmuffin.protocol.commands.lobby.training.IdentifyResponse;
import bluffinmuffin.protocol.commands.lobby.training.ListTableTrainingResponse;

public class LobbyTCPClient extends Thread
{
    private String m_playerName;
    private final String m_serverAddress;
    private final int m_serverPort;
    
    // List of ClientSidePokerTcpServer (one for each table the player joined)
    private final Map<Integer, GameTCPClient> m_clients = new HashMap<Integer, GameTCPClient>();
    private final BlockingQueue<String> m_incoming = new LinkedBlockingQueue<String>();
    
    // Communication with the server
    Socket m_socket = null;
    BufferedReader m_fromServer = null;
    PrintWriter m_toServer = null;
    
    public LobbyTCPClient(String serverAddress, int serverPort)
    {
        m_serverAddress = serverAddress;
        m_serverPort = serverPort;
    }
    
    public boolean connect()
    {
        try
        {
            m_socket = new Socket(m_serverAddress, m_serverPort);
            m_toServer = new PrintWriter(m_socket.getOutputStream(), true); // Auto-flush
            m_fromServer = new BufferedReader(new InputStreamReader(m_socket.getInputStream()));
        }
        catch (final Exception e)
        {
            System.err.println("Error on connect: " + e.getMessage());
            return false;
        }
        return true;
    }
    
    private StringTokenizer receiveCommand(String expected)
    {
        String s = null;
        try
        {
            s = m_incoming.take();
        }
        catch (final InterruptedException e1)
        {
            e1.printStackTrace();
        }
        StringTokenizer token = new StringTokenizer(s, AbstractLobbyCommand.Delimitter + "");
        String commandName = token.nextToken();
        while (s == null || !commandName.equals(expected))
        {
            try
            {
                s = m_incoming.take();
            }
            catch (final InterruptedException e)
            {
                e.printStackTrace();
            }
            token = new StringTokenizer(s, AbstractLobbyCommand.Delimitter + "");
            commandName = token.nextToken();
        }
        return token;
    }
    
    public boolean identify(String name)
    {
        m_playerName = name;
        send(new IdentifyCommand(m_playerName));
        final StringTokenizer token = receiveCommand(IdentifyResponse.COMMAND_NAME);
        final IdentifyResponse response = new IdentifyResponse(token);
        return response.isOK();
    }
    
    public boolean isConnected()
    {
        return (m_socket != null) && m_socket.isConnected() && !m_socket.isClosed();
    }
    
    public void disconnect()
    {
        try
        {
            // Disconnect all clients (PokerClient).
            while (m_clients.size() != 0)
            {
                m_clients.get(0).disconnect();
                m_clients.remove(0);
            }
            if (isConnected())
            {
                // Alors on disconnect
                send(new DisconnectCommand());
                m_fromServer.close();
                m_toServer.close();
                m_socket.close();
                m_socket = null;
                m_fromServer = null;
                m_toServer = null;
            }
        }
        catch (final IOException e)
        {
            e.printStackTrace();
        }
    }
    
    public void sendMessage(String p_msg)
    {
        System.out.println(m_playerName + " SEND [" + p_msg + "]");
        m_toServer.println(p_msg);
    }
    
    public void send(AbstractCommand p_msg)
    {
        sendMessage(p_msg.encode());
    }
    
    public String getPlayerName()
    {
        return m_playerName;
    }
    
    public String getServerAddress()
    {
        return m_serverAddress;
    }
    
    public int getServerPort()
    {
        return m_serverPort;
    }
    
    public GameTCPClient findClient(int noPort)
    {
        return m_clients.get(noPort);
    }
    
    public GameTCPClient joinTable(int p_noPort, String p_tableName, IPokerViewer gui)
    {
        // Build query.
        final JoinTableCommand command = new JoinTableCommand(m_playerName, p_noPort);
        
        // Send query.
        m_toServer.println(command.encode());
        
        // Wait for response.
        final StringTokenizer token2 = receiveCommand(JoinTableResponse.COMMAND_NAME);
        final JoinTableResponse response2 = new JoinTableResponse(token2);
        final int noSeat = response2.getNoSeat();
        
        if (noSeat == -1)
        {
            System.out.println("Cannot sit at this table: " + p_tableName);
            return null;
        }
        
        final GameTCPClient client = new GameTCPClient(this, p_noPort, noSeat, m_playerName);
        m_clients.put(p_noPort, client);
        if (gui != null)
        {
            gui.setGame(client, client.getNoSeat());
            gui.start();
        }
        client.start();
        return client;
    }
    
    public int createTable(String p_tableName, int p_bigBlind, int p_maxPlayers, int wtaPlayerAction, int wtaBoardDealed, int wtaPotWon, GameBetLimitType limit, int startingMoney)
    {
        // Send query.
        send(new CreateTrainingTableCommand(p_tableName, p_bigBlind, p_maxPlayers, m_playerName, wtaPlayerAction, wtaBoardDealed, wtaPotWon, limit, startingMoney));
        // Wait for response.
        
        final StringTokenizer token = receiveCommand(CreateTrainingTableResponse.COMMAND_NAME);
        final CreateTrainingTableResponse response = new CreateTrainingTableResponse(token);
        return response.getResponsePort();
    }
    
    public List<TupleTableInfoCareer> getListTablesCareer()
    {
        // Ask the server for all available tables.
        send(new ListTableCommand(false));
        
        final StringTokenizer token = receiveCommand(ListTableCareerResponse.COMMAND_NAME);
        final ListTableCareerResponse response = new ListTableCareerResponse(token);
        return response.getTables();
    }
    
    public List<TupleTableInfoTraining> getListTablesTraining()
    {
        // Ask the server for all available tables.
        send(new ListTableCommand(true));
        
        final StringTokenizer token = receiveCommand(ListTableTrainingResponse.COMMAND_NAME);
        final ListTableTrainingResponse response = new ListTableTrainingResponse(token);
        return response.getTables();
    }
    
    @Override
    public void run()
    {
        try
        {
            while (isConnected())
            {
                receive();
            }
        }
        catch (final IOException e)
        {
            return;
        }
        catch (final InterruptedException e)
        {
            e.printStackTrace();
        }
    }
    
    private void receive() throws IOException, InterruptedException
    {
        System.out.println(m_playerName + " IS WAITING");
        final String line = m_fromServer.readLine();
        System.out.println(m_playerName + " RECV [" + line + "]");
        final StringTokenizer token = new StringTokenizer(line, AbstractLobbyCommand.Delimitter + "");
        final String commandName = token.nextToken();
        if (commandName.equals(GameCommand.COMMAND_NAME))
        {
            final GameCommand c = new GameCommand(token);
            while (!m_clients.containsKey(c.getTableId()))
            {
                Thread.sleep(100);
            }
            
            m_clients.get(c.getTableId()).incoming(c.getCommand());
        }
        else
        {
            m_incoming.put(line);
        }
    }
}
